sql: Only Unicode NFC normalize SQL names if necessary

nvb opened the pull request:
`NormalizeName` is currently the leading source of total allocations
(~7% of total allocations for a single-node under photos load). The function
consists of mapping the provided name to lower case, then Unicode NFC
normalizing the name. `strings.Map` will only allocate when it makes
changes, but `norm.NFC.String` allocates on all calls. This change
uses the `norm.NFC.IsNormalString` to avoid Unicode normalizing when
it isn't necessary.

Interestingly, the change seems to have increased allocation count, but
it also had a small speed performance improvement across the board.
Therefore, I'm not positive we actually want to merge the change. Still,
I'm hoping this will at least bring to light that `NormalizeName` is not
free, so we may want to consider saving/memoizing the result in cases
where it is called multiple times for the same name.

```
name                                old time/op    new time/op    delta
Bank2_Cockroach-2                    3.38ms ±268%    1.16ms ±12%    ~            (p=0.083 n=10+8)
Bank4_Cockroach-2                     1.19ms ± 7%    1.19ms ±10%    ~             (p=0.815 n=9+8)
Bank8_Cockroach-2                     1.16ms ± 3%    1.14ms ± 5%    ~           (p=0.190 n=10+10)
Bank16_Cockroach-2                    1.07ms ± 4%    1.08ms ± 6%    ~           (p=0.579 n=10+10)
Bank32_Cockroach-2                    1.01ms ± 6%    1.00ms ± 5%    ~            (p=0.447 n=9+10)
Bank64_Cockroach-2                     934µs ± 4%     931µs ± 6%    ~           (p=1.000 n=10+10)
Select1_Cockroach-2                   46.8µs ± 2%    46.2µs ± 1%  -1.20%         (p=0.017 n=10+9)
Select2_Cockroach-2                    856µs ± 2%     827µs ± 1%  -3.38%         (p=0.000 n=10+9)
Insert1_Cockroach-2                    451µs ± 1%     447µs ± 2%  -1.02%        (p=0.009 n=10+10)
Insert10_Cockroach-2                   802µs ± 1%     794µs ± 1%  -1.03%         (p=0.000 n=10+8)
Insert100_Cockroach-2                 4.13ms ± 2%    4.08ms ± 1%  -1.11%        (p=0.019 n=10+10)
Insert1000_Cockroach-2                40.0ms ± 1%    39.7ms ± 1%    ~           (p=0.190 n=10+10)
Update1_Cockroach-2                    653µs ± 2%     651µs ± 2%    ~           (p=0.529 n=10+10)
Update10_Cockroach-2                  1.33ms ± 1%    1.31ms ± 2%  -1.08%        (p=0.015 n=10+10)
Update100_Cockroach-2                 7.37ms ± 1%    7.19ms ± 1%  -2.44%        (p=0.000 n=10+10)
Update1000_Cockroach-2                65.9ms ± 3%    65.9ms ± 4%    ~           (p=0.971 n=10+10)
Delete1_Cockroach-2                    642µs ± 2%     640µs ± 2%    ~           (p=0.579 n=10+10)
Delete10_Cockroach-2                  2.00ms ± 1%    1.97ms ± 1%  -1.48%        (p=0.002 n=10+10)
Delete100_Cockroach-2                 16.6ms ± 2%    16.3ms ± 2%  -1.93%        (p=0.000 n=10+10)
Delete1000_Cockroach-2                 175ms ± 2%     173ms ± 1%    ~           (p=0.052 n=10+10)
Scan1_Cockroach-2                      170µs ± 2%     169µs ± 2%    ~            (p=0.315 n=8+10)
Scan10_Cockroach-2                     213µs ± 2%     212µs ± 2%    ~             (p=0.481 n=9+8)
Scan100_Cockroach-2                    589µs ± 1%     560µs ± 3%  -4.95%         (p=0.000 n=9+10)
Scan1000_Cockroach-2                  4.29ms ± 2%    3.99ms ± 1%  -7.17%        (p=0.000 n=10+10)
Scan10000_Cockroach-2                 44.5ms ± 0%    41.3ms ± 3%  -7.11%         (p=0.000 n=9+10)
Scan1000Limit1_Cockroach-2             182µs ± 1%     179µs ± 1%  -1.20%          (p=0.001 n=9+9)
Scan1000Limit10_Cockroach-2            223µs ± 1%     223µs ± 1%    ~           (p=0.684 n=10+10)
Scan1000Limit100_Cockroach-2           610µs ± 4%     586µs ± 3%  -4.01%        (p=0.000 n=10+10)
Scan10000FilterLimit1_Cockroach-2      290µs ± 2%     287µs ± 2%  -0.87%         (p=0.028 n=10+9)
Scan10000FilterLimit10_Cockroach-2     349µs ± 2%     344µs ± 1%  -1.25%        (p=0.004 n=10+10)
Scan10000FilterLimit50_Cockroach-2    14.5ms ± 1%    13.3ms ± 1%  -8.13%         (p=0.000 n=9+10)
PgbenchQuery_Cockroach-2              3.15ms ± 3%    3.20ms ± 8%    ~            (p=0.536 n=7+10)

name                                old alloc/op   new alloc/op   delta
Bank2_Cockroach-2                     95.9kB ± 2%    96.0kB ± 2%    ~           (p=0.853 n=10+10)
Bank4_Cockroach-2                     96.7kB ± 1%    97.3kB ± 1%  +0.53%        (p=0.029 n=10+10)
Bank8_Cockroach-2                     96.1kB ± 1%    96.6kB ± 1%    ~           (p=0.089 n=10+10)
Bank16_Cockroach-2                    95.0kB ± 2%    95.4kB ± 1%    ~           (p=0.143 n=10+10)
Bank32_Cockroach-2                    93.5kB ± 1%    94.1kB ± 1%  +0.62%        (p=0.043 n=10+10)
Bank64_Cockroach-2                    92.6kB ± 0%    93.0kB ± 0%  +0.39%        (p=0.015 n=10+10)
Select1_Cockroach-2                   2.27kB ± 0%    2.27kB ± 0%    ~            (p=0.134 n=8+10)
Select2_Cockroach-2                   72.9kB ± 0%    73.8kB ± 0%  +1.20%         (p=0.000 n=9+10)
Insert1_Cockroach-2                   23.7kB ± 0%    23.7kB ± 0%    ~            (p=0.473 n=8+10)
Insert10_Cockroach-2                  64.2kB ± 0%    64.3kB ± 0%  +0.14%         (p=0.002 n=10+9)
Insert100_Cockroach-2                  465kB ± 0%     465kB ± 0%    ~            (p=0.315 n=9+10)
Insert1000_Cockroach-2                4.20MB ± 2%    4.19MB ± 1%    ~           (p=0.353 n=10+10)
Update1_Cockroach-2                   36.4kB ± 0%    36.5kB ± 0%  +0.36%        (p=0.000 n=10+10)
Update10_Cockroach-2                   102kB ± 0%     102kB ± 0%    ~            (p=0.515 n=10+8)
Update100_Cockroach-2                  717kB ± 0%     717kB ± 0%    ~           (p=0.739 n=10+10)
Update1000_Cockroach-2                6.26MB ± 1%    6.28MB ± 1%    ~           (p=0.631 n=10+10)
Delete1_Cockroach-2                   30.2kB ± 0%    30.6kB ± 0%  +1.16%         (p=0.000 n=10+9)
Delete10_Cockroach-2                  72.8kB ± 0%    73.2kB ± 0%  +0.52%         (p=0.000 n=10+9)
Delete100_Cockroach-2                  494kB ± 0%     494kB ± 0%    ~            (p=0.113 n=10+9)
Delete1000_Cockroach-2                4.41MB ± 0%    4.43MB ± 1%    ~           (p=0.579 n=10+10)
Scan1_Cockroach-2                     11.3kB ± 0%    11.3kB ± 0%    ~            (p=0.795 n=10+9)
Scan10_Cockroach-2                    15.2kB ± 0%    15.2kB ± 0%    ~            (p=0.484 n=9+10)
Scan100_Cockroach-2                   47.4kB ± 0%    47.4kB ± 0%    ~           (p=0.425 n=10+10)
Scan1000_Cockroach-2                   336kB ± 0%     336kB ± 0%  -0.01%          (p=0.018 n=9+9)
Scan10000_Cockroach-2                 5.79MB ± 0%    5.79MB ± 0%  -0.01%        (p=0.000 n=10+10)
Scan1000Limit1_Cockroach-2            11.9kB ± 0%    11.9kB ± 0%    ~           (p=0.210 n=10+10)
Scan1000Limit10_Cockroach-2           15.7kB ± 0%    15.7kB ± 0%    ~             (p=0.863 n=9+9)
Scan1000Limit100_Cockroach-2          47.8kB ± 0%    47.9kB ± 0%  +0.03%        (p=0.050 n=10+10)
Scan10000FilterLimit1_Cockroach-2     24.0kB ± 0%    24.0kB ± 0%  +0.30%        (p=0.000 n=10+10)
Scan10000FilterLimit10_Cockroach-2    28.5kB ± 0%    28.6kB ± 0%  +0.24%        (p=0.000 n=10+10)
Scan10000FilterLimit50_Cockroach-2    1.29MB ± 0%    1.29MB ± 0%  +0.01%        (p=0.002 n=10+10)
PgbenchQuery_Cockroach-2               203kB ± 5%     209kB ±10%    ~            (p=0.055 n=7+10)

name                                old allocs/op  new allocs/op  delta
Bank2_Cockroach-2                      1.96k ± 2%     2.02k ± 2%  +2.79%        (p=0.000 n=10+10)
Bank4_Cockroach-2                      1.98k ± 1%     2.05k ± 1%  +3.21%         (p=0.000 n=10+9)
Bank8_Cockroach-2                      1.97k ± 1%     2.04k ± 1%  +3.19%        (p=0.000 n=10+10)
Bank16_Cockroach-2                     1.95k ± 1%     2.01k ± 1%  +3.13%        (p=0.000 n=10+10)
Bank32_Cockroach-2                     1.93k ± 1%     1.99k ± 1%  +3.39%        (p=0.000 n=10+10)
Bank64_Cockroach-2                     1.91k ± 0%     1.97k ± 1%  +3.14%        (p=0.000 n=10+10)
Select1_Cockroach-2                     50.0 ± 0%      50.0 ± 0%    ~     (all samples are equal)
Select2_Cockroach-2                    2.34k ± 0%     2.44k ± 0%  +4.35%         (p=0.000 n=6+10)
Insert1_Cockroach-2                      331 ± 0%       333 ± 0%  +0.45%         (p=0.001 n=9+10)
Insert10_Cockroach-2                     718 ± 0%       720 ± 0%  +0.25%         (p=0.002 n=6+10)
Insert100_Cockroach-2                  4.45k ± 0%     4.45k ± 0%    ~            (p=0.343 n=9+10)
Insert1000_Cockroach-2                 41.9k ± 1%     41.9k ± 0%    ~           (p=0.393 n=10+10)
Update1_Cockroach-2                      636 ± 0%       654 ± 0%  +2.80%         (p=0.000 n=9+10)
Update10_Cockroach-2                   1.26k ± 0%     1.27k ± 0%  +1.27%         (p=0.000 n=10+6)
Update100_Cockroach-2                  7.16k ± 0%     7.17k ± 0%  +0.25%         (p=0.000 n=9+10)
Update1000_Cockroach-2                 63.5k ± 0%     63.6k ± 1%  +0.15%        (p=0.022 n=10+10)
Delete1_Cockroach-2                      525 ± 0%       569 ± 0%  +8.38%          (p=0.000 n=7+8)
Delete10_Cockroach-2                   1.11k ± 0%     1.16k ± 0%  +3.99%         (p=0.000 n=8+10)
Delete100_Cockroach-2                  6.82k ± 0%     6.86k ± 0%  +0.64%         (p=0.000 n=10+9)
Delete1000_Cockroach-2                 63.6k ± 0%     63.7k ± 0%  +0.21%        (p=0.000 n=10+10)
Scan1_Cockroach-2                        215 ± 0%       216 ± 0%  +0.33%         (p=0.012 n=10+7)
Scan10_Cockroach-2                       295 ± 0%       297 ± 0%  +0.54%         (p=0.000 n=7+10)
Scan100_Cockroach-2                    1.02k ± 0%     1.02k ± 0%  +0.07%        (p=0.019 n=10+10)
Scan1000_Cockroach-2                   8.23k ± 0%     8.23k ± 0%  +0.01%          (p=0.000 n=9+9)
Scan10000_Cockroach-2                  80.4k ± 0%     80.4k ± 0%  +0.00%         (p=0.018 n=9+10)
Scan1000Limit1_Cockroach-2               232 ± 0%       233 ± 0%  +0.26%         (p=0.015 n=10+9)
Scan1000Limit10_Cockroach-2              311 ± 0%       312 ± 0%  +0.32%          (p=0.000 n=9+9)
Scan1000Limit100_Cockroach-2           1.04k ± 0%     1.04k ± 0%  +0.10%        (p=0.001 n=10+10)
Scan10000FilterLimit1_Cockroach-2        496 ± 0%       505 ± 0%  +1.79%        (p=0.000 n=10+10)
Scan10000FilterLimit10_Cockroach-2       656 ± 0%       665 ± 0%  +1.39%        (p=0.000 n=10+10)
Scan10000FilterLimit50_Cockroach-2     27.1k ± 0%     27.1k ± 0%  +0.03%        (p=0.000 n=10+10)
PgbenchQuery_Cockroach-2               3.22k ± 3%     3.40k ± 7%    ~            (p=0.053 n=7+10)
```
## EDIT

This now only limits the number of calls to the function. We had a few instances where we were calling it in loops unnecessarily.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5370)

<!-- Reviewable:end -->

tschottdorf commented on sql/keys.go:
```diff
@@ -57,7 +57,11 @@ var normalize = unicode.SpecialCase{
 // (NFC).
 func NormalizeName(name string) string {
-	return norm.NFC.String(strings.Map(normalize.ToLower, name))
+	lower := strings.Map(normalize.ToLower, name)
```
with a string-adapted LRU cache, wouldn't you be able to get down to mostly zero (I'm assuming that NormalizeName isn't used on user data, just on schema and the like)?

tschottdorf replied:
ah, nvm. You suggest that in the commit message.

petermattis commented on sql/keys.go:
```diff
@@ -57,7 +57,11 @@ var normalize = unicode.SpecialCase{
 func NormalizeName(name string) string {
-	return norm.NFC.String(strings.Map(normalize.ToLower, name))
+	lower := strings.Map(normalize.ToLower, name)
+	if !norm.NFC.IsNormalString(lower) {
```
`IsNormalString` isn't simple: https://github.com/golang/text/blob/master/unicode/norm/normalize.go#L106. I wonder if it would be better to check some sort of quick `IsASCII` check.

nvanbenschoten replied:
I'll play around with a few simpler checks that get us 90% of the way there.

petermattis commented:
I'm suspicious that the benchmark results are spurious. Benchmarking on laptops is prone due to the thermal CPU scaling (the CPU is throttled when it gets hotter). Would be better to get you set up with a VM in AWS or GCE to benchmark on.

nvanbenschoten commented:
I've been running benchmarks on AWS for the past few weeks, so most of those issues shouldn't be concerns. The only thing that worries me slightly is that my current approach runs all ~30 min of "after" benchmarks before running all ~30 min of "before" benchmarks. I'm curious if interleaving them would give results that are less susceptible to outside influences.

RaduBerinde commented:
I did 20 runs (alternating before and after) on my desktop last night, using:
```
for i in `seq 1 20`; do git checkout b293e8b; make bench TESTS=_Cockroach >> /work/before; rm-rf /tmp/*; git checkout pr/5370; make bench TESTS=_Cockroach >> /work/after; rm -rf /tmp/*; done
```

I only see random variation.

```
Bank2_Cockroach-4                    1.38ms ±63%   1.24ms ±33%     ~     (p=0.260 n=17+16)
Bank4_Cockroach-4                   1.44ms ±184%   1.14ms ±20%     ~     (p=0.568 n=17+18)
Bank8_Cockroach-4                   4.04ms ±228%  1.77ms ±311%     ~     (p=0.198 n=20+17)
Bank16_Cockroach-4                    891µs ±19%    871µs ±20%     ~     (p=0.533 n=17+16)
Bank32_Cockroach-4                    790µs ±33%    755µs ±28%     ~     (p=0.357 n=20+17)
Bank64_Cockroach-4                    671µs ±20%    733µs ±18%   +9.28%  (p=0.002 n=17+20)
Select1_Cockroach-4                  37.0µs ±13%   39.3µs ±38%     ~     (p=0.551 n=16+18)
Select2_Cockroach-4                   736µs ±26%    747µs ±54%     ~     (p=0.372 n=18+18)
Insert1_Cockroach-4                   463µs ±42%    428µs ±10%     ~     (p=0.251 n=18+16)
Insert10_Cockroach-4                  791µs ±32%    767µs ±24%     ~     (p=0.869 n=19+18)
Insert100_Cockroach-4                3.60ms ±26%   3.52ms ±12%     ~     (p=0.799 n=19+18)
Insert1000_Cockroach-4               34.2ms ±21%   33.9ms ±22%     ~     (p=0.370 n=19+19)
Update1_Cockroach-4                   593µs ± 2%    607µs ±10%     ~     (p=0.102 n=16+19)
Update10_Cockroach-4                 1.20ms ± 2%   1.24ms ±15%     ~     (p=0.159 n=18+19)
Update100_Cockroach-4                6.87ms ±35%   6.87ms ±20%     ~     (p=0.696 n=18+18)
Update1000_Cockroach-4               56.3ms ± 5%   56.2ms ± 7%     ~     (p=0.673 n=18+18)
Delete1_Cockroach-4                   727µs ±25%    690µs ±16%     ~     (p=0.134 n=19+20)
Delete10_Cockroach-4                 1.86ms ± 9%   1.85ms ± 8%     ~     (p=0.749 n=19+20)
Delete100_Cockroach-4                15.0ms ± 9%   15.3ms ±14%     ~     (p=0.372 n=18+18)
Delete1000_Cockroach-4                152ms ± 9%    150ms ± 8%     ~     (p=0.461 n=20+19)
Scan1_Cockroach-4                     158µs ±13%    142µs ±22%  -10.21%  (p=0.000 n=16+15)
Scan10_Cockroach-4                    207µs ±40%    196µs ±25%     ~     (p=0.480 n=19+18)
Scan100_Cockroach-4                   472µs ± 7%    521µs ±36%     ~     (p=0.144 n=16+18)
Scan1000_Cockroach-4                 3.19ms ± 3%   3.47ms ±36%   +8.93%  (p=0.011 n=17+16)
Scan10000_Cockroach-4                36.9ms ±42%   32.7ms ± 3%     ~     (p=0.066 n=19+15)
Scan1000Limit1_Cockroach-4            151µs ± 5%    147µs ± 6%   -2.68%  (p=0.003 n=17+16)
Scan1000Limit10_Cockroach-4           196µs ±18%    209µs ±46%     ~     (p=0.799 n=19+18)
Scan1000Limit100_Cockroach-4          517µs ±32%    495µs ±10%     ~     (p=0.384 n=18+16)
Scan10000FilterLimit1_Cockroach-4     248µs ±15%    250µs ±27%     ~     (p=0.198 n=16+18)
Scan10000FilterLimit10_Cockroach-4    297µs ±21%    285µs ± 8%     ~     (p=0.089 n=18+17)
Scan10000FilterLimit50_Cockroach-4   10.8ms ±16%   11.2ms ±32%     ~     (p=0.339 n=18+18)
PgbenchQuery_Cockroach-4             2.85ms ±16%   2.80ms ±18%     ~     (p=0.546 n=17+18)
```

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5370#-:-KDTZ0qBeyH_3p5FXMMX:2019628740)*
<!-- Sent from Reviewable.io -->

nvb committed:
sql: Limit the number of calls to NormalizeName

`NormalizeName` is currently the leading source of total allocations
(~7% of total allocations for a single-node under load). The function
consists of mapping the provided name to lower case, then Unicode NFC
normalizing the name. `strings.Map` will only allocate when it makes
changes, but `norm.NFC.String` allocates on all calls. This change
reduces the number of calls to the function in a number of places.

nvanbenschoten commented:
Yeah like I said, I wasnt sure we actually wanted to merge this. I just wanted to bring it to attention that `NormalizeName` isnt as lightweight as is being assumed in places like [this](https://github.com/cockroachdb/cockroach/blob/master/sql/structured.go#L398) or [this](https://github.com/cockroachdb/cockroach/blob/master/sql/structured.go#L449). For now, I'll just clean those two uses up. I've changed the commit to fix a number of places where we were unnecessarily calling `NormalizeName`.

tamird commented:
Reviewed 1 of 5 files at r3.
Review status: 1 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/sort.go, line 76 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5370#-KDjmHWpCNt3DV2R38d8:-KDjmHWpCNt3DV2R38d9:-282216393) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ca3821d58641fa35148de241e7f780aafe5a06/sql/sort.go#L76)):*
throughout: you said in the commit message "`strings.Map` will only allocate when it makes changes" - in line of that, couldn't you keep `equalName` here, or does it make a difference? (note pulling the target's NormalizeName out of the loop should obviously stay).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5370)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 1 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/sort.go, line 76 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5370#-KDjmHWpCNt3DV2R38d8:-KDjnUtPCuRWGU53TPD3:-779605573) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6ca3821d58641fa35148de241e7f780aafe5a06/sql/sort.go#L76)):*
`strings.Map` will only allocate when it makes changes, but that's not true of `NormalizeName` as a whole. `norm.NFC.String` always allocates and `isASCII` doesnt guarantee to catch all normalized names.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5370)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2, 4 of 5 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5370#-:-KDxasfX6jmUVNXPG9rH:1417088721)*
<!-- Sent from Reviewable.io -->
